src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -30,61 +29,59 @@ namespace Microsoft.IdentityModel.Tokens
         private readonly int _removeExpiredValuesIntervalInSeconds;
-        // if true, then items will be maintained in a LRU fashion, moving to front of list when accessed in the cache.
-        private readonly bool _maintainLRU;
-
-        private readonly TaskCreationOptions _options;
-        private DateTime _dueForExpiredValuesRemoval;
-
         // for testing purpose only to verify the task count
         private int _taskCount = 0;
+        private DateTime _timeForNextExpiredValuesRemoval;
+        private DateTime _timeForNextCompaction;
 
         #region event queue
-  
         private int _eventQueuePollingInterval = 50;
-
         // The idle timeout, the _eventQueueTask will end after being idle for the specified time interval (execution continues even if the queue is empty to reduce the task startup overhead), default to 120 seconds.
         // TODO: consider implementing a better algorithm that tracks and predicts the usage patterns and adjusts this value dynamically.
         private long _eventQueueTaskIdleTimeoutInSeconds = 120;
-
         // The time when the _eventQueueTask should end. The intent is to reduce the overhead costs of starting/ending tasks too frequently
         // but at the same time keep the _eventQueueTask a short running task.
         // Since Task is based on thread pool the overhead should be reasonable.
         private DateTime _eventQueueTaskStopTime;
-
         // task states used to ensure thread safety (Interlocked.CompareExchange)
         private const int EventQueueTaskStopped = 0; // task not started yet
         private const int EventQueueTaskRunning = 1; // task is running
         private const int EventQueueTaskDoNotStop = 2; // force the task to continue even it has past the _eventQueueTaskStopTime, see StartEventQueueTaskIfNotRunning() for more details.
         private int _eventQueueTaskState = EventQueueTaskStopped;
 
-        private const int CompactionNotQueued = 0; // compaction action not in the event queue
-        private const int CompactionQueuedOrRunning = 1; // compaction action in the event queue or currently in progress
-        private int _compactionState = CompactionNotQueued;
+        private const int ActionNotQueued = 0; // compaction action not in the event queue
+        private const int ActionQueuedOrRunning = 1; // compaction action in the event queue or currently in progress
+        private int _compactValuesState = ActionNotQueued;
+        private int _removeExpiredValuesState = ActionNotQueued;
+        private int _processCompactedValuesState = ActionNotQueued;
 
         // set to true when the AppDomain is to be unloaded or the default AppDomain process is ready to exit
         private bool _shouldStopImmediately = false;
 
-        internal ItemRemoved OnItemRemoved
-        {
-            get;
-            set;
-        }
+        internal ItemRemoved OnItemRemoved { get; set; }
+
+        internal ItemRemoved OnItemExpired { get; set; }
```

Comment:
Consider using a different delegate per callback, It reads odd to have ItemRemoved Delegate for for OnItemExpired,

or just use Func<TValue, bool> since these are all internal.

Reply:
@keegan-caruso Nice catch, will fix.